File: ios/iOSBaseProject/Presentation/Features/Auth/ViewModels/AuthViewModel.swift
//
//  AuthViewModel.swift
//  BaatCheet
//
//  ViewModel for Authentication - MVVM Pattern
//

import Foundation
import Combine

/// Authentication UI State
struct AuthState: Equatable {
    var isLoading: Bool = false
    var isAuthenticated: Bool = false
    var needsVerification: Bool = false
    var currentUser: User? = nil
    var pendingEmail: String? = nil
    var error: String? = nil
    
    static func == (lhs: AuthState, rhs: AuthState) -> Bool {
        lhs.isLoading == rhs.isLoading &&
        lhs.isAuthenticated == rhs.isAuthenticated &&
        lhs.needsVerification == rhs.needsVerification &&
        lhs.currentUser?.id == rhs.currentUser?.id &&
        lhs.pendingEmail == rhs.pendingEmail &&
        lhs.error == rhs.error
    }
}

/// Authentication Events
enum AuthEvent {
    case signIn(email: String, password: String)
    case signUp(email: String, password: String, firstName: String?, lastName: String?)
    case verifyEmail(code: String)
    case resendCode
    case logout
    case clearError
    case setPendingEmail(String)
}

/// Authentication ViewModel
/// Follows MVVM pattern with unidirectional data flow.
@MainActor
final class AuthViewModel: ObservableObject {
    
    // MARK: - Published State
    
    @Published private(set) var state = AuthState()
    
    // MARK: - Dependencies (Use Cases)
    
    private let signInUseCase: SignInUseCaseProtocol
    private let signUpUseCase: SignUpUseCaseProtocol
    private let verifyEmailUseCase: VerifyEmailUseCaseProtocol
    private let logoutUseCase: LogoutUseCaseProtocol
    private let getCurrentUserUseCase: GetCurrentUserUseCaseProtocol
    private let isAuthenticatedUseCase: IsAuthenticatedUseCaseProtocol
    
    // MARK: - Private State
    
    private var pendingEmail: String?
    
    // MARK: - Initialization
    
    init(
        signInUseCase: SignInUseCaseProtocol,
        signUpUseCase: SignUpUseCaseProtocol,
        verifyEmailUseCase: VerifyEmailUseCaseProtocol,
        logoutUseCase: LogoutUseCaseProtocol,
        getCurrentUserUseCase: GetCurrentUserUseCaseProtocol,
        isAuthenticatedUseCase: IsAuthenticatedUseCaseProtocol
    ) {
        self.signInUseCase = signInUseCase
        self.signUpUseCase = signUpUseCase
        self.verifyEmailUseCase = verifyEmailUseCase
        self.logoutUseCase = logoutUseCase
        self.getCurrentUserUseCase = getCurrentUserUseCase
        self.isAuthenticatedUseCase = isAuthenticatedUseCase
        
        // Check initial auth state
        checkAuthState()
    }
    
    // MARK: - Event Handling
    
    func handle(_ event: AuthEvent) {
        switch event {
        case .signIn(let email, let password):
            Task { await signIn(email: email, password: password) }
            
        case .signUp(let email, let password, let firstName, let lastName):
            Task { await signUp(email: email, password: password, firstName: firstName, lastName: lastName) }
            
        case .verifyEmail(let code):
            Task { await verifyEmail(code: code) }
            
        case .resendCode:
            Task { await resendCode() }
            
        case .logout:
            Task { await logout() }
            
        case .clearError:
            state.error = nil
            
        case .setPendingEmail(let email):
            pendingEmail = email
            state.pendingEmail = email
        }
    }
    
    // MARK: - Private Methods
    
    private func checkAuthState() {
        if isAuthenticatedUseCase.execute() {
            state.isAuthenticated = true
            state.currentUser = getCurrentUserUseCase.execute()
        }
    }
    
    private func signIn(email: String, password: String) async {
        state.isLoading = true
        state.error = nil
        
        let result = await signInUseCase.execute(email: email, password: password)
        
        handleAuthResult(result, email: email)
    }
    
    private func signUp(email: String, password: String, firstName: String?, lastName: String?) async {
        state.isLoading = true
        state.error = nil
        
        let credentials = AuthCredentials(
            email: email,
            password: password,
            firstName: firstName,
            lastName: lastName
        )
        
        let result = await signUpUseCase.execute(credentials: credentials)
        
        handleAuthResult(result, email: email)
    }
    
    private func verifyEmail(code: String) async {
        guard let email = pendingEmail ?? state.pendingEmail else {
            state.error = "No pending email for verification"
            return
        }
        
        state.isLoading = true
        state.error = nil
        
        let result = await verifyEmailUseCase.execute(email: email, code: code)
        
        switch result {
        case .success(_, _, let user):
            pendingEmail = nil
            state.isLoading = false
            state.isAuthenticated = true
            state.needsVerification = false
            state.currentUser = user
            state.pendingEmail = nil
            state.error = nil
            
        case .needsVerification:
            state.isLoading = false
            state.error = "Invalid code. Please try again."
            
        case .failure(let error):
            state.isLoading = false
            state.error = error.localizedDescription
        }
    }
    
    private func resendCode() async {
        // This would typically call a resend use case
        // For now, just clear any error
        state.error = nil
    }
    
    private func logout() async {
        state.isLoading = true
        
        await logoutUseCase.execute()
        
        // Reset state
        state = AuthState()
    }
    
    private func handleAuthResult(_ result: AuthResult, email: String) {
        switch result {
        case .success(_, _, let user):
            state.isLoading = false
            state.isAuthenticated = true
            state.currentUser = user
            state.error = nil
            
        case .needsVerification:
            pendingEmail = email
            state.isLoading = false
            state.needsVerification = true
            state.pendingEmail = email
            
        case .failure(let error):
            state.isLoading = false
            state.error = error.localizedDescription
        }
    }
}

// MARK: - Convenience Factory

extension AuthViewModel {
    /// Create AuthViewModel with default dependencies
    static func create() -> AuthViewModel {
        DependencyContainer.shared.makeAuthViewModel()
    }
}

File: android/app/src/main/java/com/baatcheet/app/ui/login/AuthViewModel.kt
package com.baatcheet.app.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.baatcheet.app.domain.model.AuthResult
import com.baatcheet.app.domain.model.User
import com.baatcheet.app.domain.usecase.auth.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Auth ViewModel
 * 
 * Follows MVVM pattern with Clean Architecture.
 * Uses StateFlow for reactive UI updates.
 * All business logic is delegated to UseCases.
 */
@HiltViewModel
class AuthViewModel @Inject constructor(
    private val signInUseCase: SignInUseCase,
    private val signUpUseCase: SignUpUseCase,
    private val verifyEmailUseCase: VerifyEmailUseCase,
    private val logoutUseCase: LogoutUseCase,
    private val getCurrentUserUseCase: GetCurrentUserUseCase,
    private val isAuthenticatedUseCase: IsAuthenticatedUseCase
) : ViewModel() {
    
    // UI State
    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state.asStateFlow()
    
    // Pending email for verification
    private var pendingEmail: String? = null
    
    init {
        // Check if user is already authenticated
        if (isAuthenticatedUseCase()) {
            _state.update { 
                it.copy(
                    isAuthenticated = true,
                    currentUser = getCurrentUserUseCase()
                ) 
            }
        }
    }
    
    /**
     * Handle UI events
     */
    fun onEvent(event: AuthEvent) {
        when (event) {
            is AuthEvent.SignIn -> signIn(event.email, event.password)
            is AuthEvent.SignUp -> signUp(event.email, event.password, event.firstName, event.lastName)
            is AuthEvent.VerifyEmail -> verifyEmail(event.code)
            is AuthEvent.ResendCode -> resendCode()
            is AuthEvent.Logout -> logout()
            is AuthEvent.ClearError -> clearError()
            is AuthEvent.SetPendingEmail -> setPendingEmail(event.email)
        }
    }
    
    private fun signIn(email: String, password: String) {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            
            when (val result = signInUseCase(email, password)) {
                is AuthResult.Success -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            isAuthenticated = true,
                            currentUser = result.user,
                            error = null
                        )
                    }
                }
                is AuthResult.NeedsVerification -> {
                    pendingEmail = email
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            needsVerification = true,
                            pendingEmail = email
                        )
                    }
                }
                is AuthResult.Failure -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            error = result.error.message ?: "Sign in failed"
                        )
                    }
                }
            }
        }
    }
    
    private fun signUp(email: String, password: String, firstName: String?, lastName: String?) {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            
            when (val result = signUpUseCase(email, password, firstName, lastName)) {
                is AuthResult.Success -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            isAuthenticated = true,
                            currentUser = result.user,
                            error = null
                        )
                    }
                }
                is AuthResult.NeedsVerification -> {
                    pendingEmail = email
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            needsVerification = true,
                            pendingEmail = email
                        )
                    }
                }
                is AuthResult.Failure -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            error = result.error.message ?: "Sign up failed"
                        )
                    }
                }
            }
        }
    }
    
    private fun verifyEmail(code: String) {
        val email = pendingEmail ?: _state.value.pendingEmail ?: return
        
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, error = null) }
            
            when (val result = verifyEmailUseCase(email, code)) {
                is AuthResult.Success -> {
                    pendingEmail = null
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            isAuthenticated = true,
                            needsVerification = false,
                            currentUser = result.user,
                            pendingEmail = null,
                            error = null
                        )
                    }
                }
                is AuthResult.NeedsVerification -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            error = "Invalid code. Please try again."
                        )
                    }
                }
                is AuthResult.Failure -> {
                    _state.update { 
                        it.copy(
                            isLoading = false,
                            error = result.error.message ?: "Verification failed"
                        )
                    }
                }
            }
        }
    }
    
    private fun resendCode() {
        // Note: This would need repository method to be exposed
        // For now, reset the countdown
        _state.update { it.copy(error = null) }
    }
    
    private fun logout() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true) }
            logoutUseCase()
            _state.update { 
                AuthState() // Reset to initial state
            }
        }
    }
    
    private fun clearError() {
        _state.update { it.copy(error = null) }
    }
    
    private fun setPendingEmail(email: String) {
        pendingEmail = email
        _state.update { it.copy(pendingEmail = email) }
    }
}

/**
 * Auth UI State
 */
data class AuthState(
    val isLoading: Boolean = false,
    val isAuthenticated: Boolean = false,
    val needsVerification: Boolean = false,
    val currentUser: User? = null,
    val pendingEmail: String? = null,
    val error: String? = null
)

/**
 * Auth UI Events
 */
sealed class AuthEvent {
    data class SignIn(val email: String, val password: String) : AuthEvent()
    data class SignUp(
        val email: String,
        val password: String,
        val firstName: String? = null,
        val lastName: String? = null
    ) : AuthEvent()
    data class VerifyEmail(val code: String) : AuthEvent()
    data object ResendCode : AuthEvent()
    data object Logout : AuthEvent()
    data object ClearError : AuthEvent()
    data class SetPendingEmail(val email: String) : AuthEvent()
}
